server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/MinorCompactor.java

```diff
@@ -148,6 +149,13 @@ public CompactionStats call() {
           reportedProblem = true;
         } catch (CompactionCanceledException e) {
           throw new IllegalStateException(e);
+        } catch (Throwable t) {
```

Comment:
```suggestion
        } catch (Error t) {
```

Catching throwable will mask checked exceptions in future code changes.  For example if the checked exceptions on a method called here are changed, and those checked exceptions should be handled.. then catching Throwable would hide the need to handle those.

Reply:
Throwable was intentional, but reducing it to an Error would be fine I guess.  My take was if a "checked" exception is thrown down the road that is not actually caught, then the system will end up in the same defunct state.

Reply:
I was just thinking if this code catches Error and and a checked exception is added elsewhere then this code will not compile the developer will have to decided what to do.

Reply:
Now I am thinking that a Halt just does not feel right, and I fear that given the appropriate circumstances could result in a cascade of tserver deaths.  I am reworking this to retry failed minor compactions in the thread that triggers minor compactions.

Reply:
@keith-turner @milleruntime Please let me know what you think about this approach: compactions that do not successfully completed get marked as failed, and then the minor compaction initiator thread will pick those up and rerun them.  Thoughts?

Reply:
> Now I am thinking that a Halt just does not feel right, and I fear that given the appropriate circumstances could result in a cascade of tserver deaths. 

That could happen.  Part of the problem is that some errors are benign and others are likely and indication of a catastrophe.  Maybe configuration is the best option could have three configuration items.

 * Configurable list of errors to retry on.
 * Configurable list of errors to halt on.
 * Action to take for errors that fall in neither list : hang, retry, halt. 

This would allow when a new error is encountered like in #1644 that the error class could just be added to config for retry.

>  It could be that all of these changes go away and I simply catch the NoClassDefFoundError and call it a day.

Maybe a follow on issue would be best.  I feel like this is a wider issue, because any thread could encounter an error and having a single mechanism for handling unexpected errors in server processes seems useful.

I am also wondering if emitting metrics for errors would be useful.  If the cardinality is deemed low enough, could emit metrics for each error class name.  Then in the metrics system could see when a tserver has an out of memory error.  This would be a follow on issue if it seems like something that might be useful.

Reply:
> Maybe a follow on issue would be best. I feel like this is a wider issue, because any thread could encounter an error and having a single mechanism for handling unexpected errors in server processes seems useful.

I agree.  I think since its clear that we need to retry flushes on the NoClassDefFoundError that should be fixed here, with anything further being follow on work.

Reply:
Thank you for your observations.  I will take out all of the retry stuff and simply handle the condition that this ticket warrants.  A more general and generic way of handling critical thread failures is needed if we want.  This one was kind of weird given the TabletMemory holds onto the memory being compacted at the same time a new memory block is used ongoing mutations so there are multiple states to handle at the same time.